src/DistributedComputations/transposable_field.jl

```diff
@@ -0,0 +1,177 @@
+- `FT`: The element type of the field. Defaults to the element type of `field_in`.
+- `with_halos`: A boolean indicating whether to include halos in the field. Defaults to `false`.
+"""
+function TransposableField(field_in, FT = eltype(field_in); with_halos = false)
+
+    zgrid = field_in.grid # We support only a 2D partition in X and Y
+    ygrid = twin_grid(zgrid; free_dimension = :y)
+    xgrid = twin_grid(zgrid; free_dimension = :x)
+
+    Nx = size(xgrid)
+    Ny = size(ygrid)
+    Nz = size(zgrid)
+
+    zarch = architecture(zgrid)
+    yarch = architecture(ygrid)
+
+    loc = location(field_in)
+
+    Rx, Ry, _ = zarch.ranks
+    if with_halos 
+        zfield = Field(loc, zgrid, FT)
+        yfield = Ry == 1 ? zfield : Field(loc, ygrid, FT)
+        xfield = Rx == 1 ? yfield : Field(loc, xgrid, FT)
+    else
+        zfield = Field(loc, zgrid, FT; indices = (1:Nz[1], 1:Nz[2], 1:Nz[3]))
+        yfield = Ry == 1 ? zfield : Field(loc, ygrid, FT; indices = (1:Ny[1], 1:Ny[2], 1:Ny[3]))
+        xfield = Rx == 1 ? yfield : Field(loc, xgrid, FT; indices = (1:Nx[1], 1:Nx[2], 1:Nx[3]))
+    end
+
+    yzbuffer = Ry == 1 ? nothing : (send = on_architecture(zarch, zeros(FT, prod(Ny))), 
+                                    recv = on_architecture(zarch, zeros(FT, prod(Nz))))
+    xybuffer = Rx == 1 ? nothing : (send = on_architecture(zarch, zeros(FT, prod(Nx))), 
+                                    recv = on_architecture(zarch, zeros(FT, prod(Ny))))
+
+    yzcomm = MPI.Comm_split(MPI.COMM_WORLD, zarch.local_index[1], zarch.local_index[1])
+    xycomm = MPI.Comm_split(MPI.COMM_WORLD, yarch.local_index[3], yarch.local_index[3])
+
+    zRx, zRy, zRz = ranks(zarch) 
+    yRx, yRy, yRz = ranks(yarch) 
+
+    yzcounts = zeros(Int, zRy * zRz)
+    xycounts = zeros(Int, yRx * yRy)
+
+    yzrank = MPI.Comm_rank(yzcomm)
+    xyrank = MPI.Comm_rank(xycomm)
+
+    yzcounts[yzrank + 1] = Ny[1] * Nz[2] * Ny[3]
+    xycounts[xyrank + 1] = Ny[1] * Nx[2] * Nx[3]
+
+    MPI.Allreduce!(yzcounts, +, yzcomm)
+    MPI.Allreduce!(xycounts, +, xycomm)
+
+    return TransposableField(xfield, yfield, zfield, 
+                             yzbuffer, xybuffer,
+                             (; yz = yzcounts, xy = xycounts),
+                             (; yz = yzcomm,   xy = xycomm))
+end
+
+#####
+##### Twin transposed grid
+#####
+
+"""
+    twin_grid(grid::DistributedGrid; free_dimension = :y)
+
+Construct a "twin" grid based on the provided distributed `grid` object.
+The twin grid is a grid that discretizes the same domain of the original grid, just with a
+different partitioning strategy  whereas the "free dimension" (i.e. the non-partitioned dimension)
+is specified by the keyword argument `free_dimension`. This could be either `:x` or `:y`.
+
+Note that `free_dimension = :z` will return the original grid as we do not allow partitioning in
+the `z` direction.
+"""
+function twin_grid(grid::DistributedGrid; free_dimension = :y)
```

Comment:
@glwagner, in another comment you asked what a `twin_grid` was. This is it!

In the code that I used to work with we called these `penciled_grid`s, which comes from the literature calling it a "pencil decomposition" when a 3D grid is partitioned along 2 different directions. Personally I find that name pretty intuitive, in case we're looking for a different name.

Reply:
I like the name twin grid. It is the same domain and discretization, just partitioned in another way. I think twin grid makes sense